File: src/ikj/main/ioke/lang/Tuple.java
/*
 * See LICENSE file in distribution for copyright and licensing information.
 */
package ioke.lang;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

import ioke.lang.exceptions.ControlFlow;

/**
 *
 * @author <a href="mailto:ola.bini@gmail.com">Ola Bini</a>
 */
public class Tuple extends IokeData {
    private final Object[] elements;

    public Tuple(Object[] elements) {
        this.elements = elements;
    }

    public static Object[] getElements(Object o) {
        return ((Tuple)IokeObject.data(o)).elements;
    }

    @Override
    public void init(IokeObject obj) throws ControlFlow {
        final Runtime runtime = obj.runtime;
        obj.setKind("Tuple");
        obj.singleMimicsWithoutCheck(runtime.origin);
        runtime.iokeGround.registerCell("Tuple", obj);

        obj.registerMethod(runtime.newNativeMethod("will modify the tuple, initializing it to contain the specified arguments", new NativeMethod("private:initializeWith") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRest("values")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    Object[] o = new Object[args.size()];
                    IokeObject.as(on, context).setData(new Tuple(args.toArray(o)));

                    return on;
                }
            }));

        obj.registerMethod(runtime.newNativeMethod("returns a new method that can be used to access an element of a tuple based on the index", new NativeMethod("private:accessor") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("index")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());
                    final int index = Number.extractInt(args.get(0), message, context);
                    return runtime.newNativeMethod("Returns the object at index " + index + " in the receiving tuple", new NativeMethod.WithNoArguments("_" + index) {
                            @Override
                            public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                                getArguments().getEvaluatedArguments(context, message, on, new ArrayList<Object>(), new HashMap<String, Object>());
                                return ((Tuple)IokeObject.data(on)).elements[index];
                            }
                        });
                }
            }));

        obj.registerMethod(runtime.newNativeMethod("returns true if the left hand side tuple is equal to the right hand side tuple.", new TypeCheckingNativeMethod("==") {
                private final TypeCheckingArgumentsDefinition ARGUMENTS = TypeCheckingArgumentsDefinition
                    .builder()
                    .receiverMustMimic(runtime.tuple)
                    .withRequiredPositional("other")
                    .getArguments();

                @Override
                public TypeCheckingArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject self, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());
                    Tuple d = (Tuple)IokeObject.data(on);
                    Object other = args.get(0);
                    boolean notResult = false;
                    if((other instanceof IokeObject) &&
                       (IokeObject.data(other) instanceof Tuple)) {
                        Tuple d2 = (Tuple)IokeObject.data(other);
                        int len = d.elements.length;
                        if(len == d2.elements.length) {
                            for(int i=0; i<len; i++) {
                                if(!d.elements[i].equals(d2.elements[i])) {
                                    notResult = true;
                                    break;
                                }
                            }
                        } else {
                            notResult = true;
                        }
                    } else {
                        notResult = true;
                    }

                    return !notResult ? context.runtime._true : context.runtime._false;
                }
            }));

        obj.registerMethod(runtime.newNativeMethod("Compares this object against the argument. The comparison is only based on the elements inside the tuple, which are in turn compared using <=>.", new TypeCheckingNativeMethod("<=>") {
                private final TypeCheckingArgumentsDefinition ARGUMENTS = TypeCheckingArgumentsDefinition
                    .builder()
                    .receiverMustMimic(runtime.tuple)
                    .withRequiredPositional("other")
                    .getArguments();

                @Override
                public TypeCheckingArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject self, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    Object arg = args.get(0);
                    Object[] one = ((Tuple)IokeObject.data(on)).elements;
                    Object[] two = ((Tuple)IokeObject.data(arg)).elements;

                    int len = Math.min(one.length, two.length);
                    SpaceshipComparator sc = new SpaceshipComparator(context, message);

                    for(int i = 0; i < len; i++) {
                        int v = sc.compare(one[i], two[i]);
                        if(v != 0) {
                            return context.runtime.newNumber(v);
                        }
                    }

                    len = one.length - two.length;

                    if(len == 0) return context.runtime.newNumber(0);
                    if(len > 0) return context.runtime.newNumber(1);
                    return context.runtime.newNumber(-1);
                }
            }));

        obj.registerMethod(runtime.newNativeMethod("Returns a text inspection of the object", new TypeCheckingNativeMethod.WithNoArguments("inspect", runtime.tuple) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                  return method.runtime.newText(Tuple.getInspect(on));
                }
            }));

        obj.registerMethod(runtime.newNativeMethod("Returns a brief text inspection of the object", new TypeCheckingNativeMethod.WithNoArguments("notice", runtime.tuple) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return method.runtime.newText(Tuple.getNotice(on));
                }
            }));

        obj.registerMethod(runtime.newNativeMethod("Returns the arity of this tuple", new TypeCheckingNativeMethod.WithNoArguments("arity", runtime.tuple) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return method.runtime.newNumber(((Tuple)IokeObject.data(on)).elements.length);
                }
            }));

        obj.registerMethod(runtime.newNativeMethod("Returns the a list representation of this tuple", new TypeCheckingNativeMethod.WithNoArguments("asList", runtime.tuple) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    return method.runtime.newList(new ArrayList<Object>(java.util.Arrays.asList(((Tuple)IokeObject.data(on)).elements)));
                }
            }));
    }

    public static String getInspect(Object on) throws ControlFlow {
        return ((Tuple)(IokeObject.data(on))).inspect(on);
    }

    public static String getNotice(Object on) throws ControlFlow {
        return ((Tuple)(IokeObject.data(on))).notice(on);
    }

    public String inspect(Object obj) throws ControlFlow {
        StringBuilder sb = new StringBuilder();
        sb.append("(");
        String sep = "";
        for(Object o : elements) {
            sb.append(sep).append(IokeObject.inspect(o));
            sep = ", ";
        }
        sb.append(")");
        return sb.toString();
    }

    public String notice(Object obj) throws ControlFlow {
        StringBuilder sb = new StringBuilder();
        sb.append("(");
        String sep = "";
        for(Object o : elements) {
            sb.append(sep).append(IokeObject.notice(o));
            sep = ", ";
        }
        sb.append(")");
        return sb.toString();
    }
}

File: src/ikc/main/Ioke.Lang/Tuple.cs
namespace Ioke.Lang {
    using System;
    using System.Collections;
    using System.Text;

    using Ioke.Lang.Util;

    public class Tuple : IokeData {
        object[] elements;

        public Tuple(object[] elements) {
            this.elements = elements;
        }

        public override void Init(IokeObject obj) {
            Runtime runtime = obj.runtime;
            obj.Kind = "Tuple";
            obj.MimicsWithoutCheck(runtime.Origin);
            runtime.IokeGround.RegisterCell("Tuple", obj);


            obj.RegisterMethod(runtime.NewNativeMethod("will modify the tuple, initializing it to contain the specified arguments",
                                                       new TypeCheckingNativeMethod("private:initializeWith", TypeCheckingArgumentsDefinition.builder()
                                                                                    .ReceiverMustMimic(runtime.Tuple)
                                                                                    .WithRest("values")
                                                                                    .Arguments,
                                                                                    (method, on, args, keywords, context, message) => {
                                                                                        IokeObject.As(on, context).Data = new Tuple(((ArrayList)args).ToArray());
                                                                                        return on;
                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("returns a new method that can be used to access an element of a tuple based on the index",
                                                       new TypeCheckingNativeMethod("private:accessor", TypeCheckingArgumentsDefinition.builder()
                                                                                    .ReceiverMustMimic(runtime.Tuple)
                                                                                    .WithRequiredPositional("index")
                                                                                    .Arguments,
                                                                                    (method, on, args, keywords, context, message) => {
                                                                                        int index = Number.ExtractInt(args[0], message, context);
                                                                                        return runtime.NewNativeMethod("Returns the object at index " + index + " in the receiving tuple",
                                                                                                                       new TypeCheckingNativeMethod.WithNoArguments("_" + index, runtime.Tuple,
                                                                                                                                                                    (method2, on2, args2, keywords2, context2, message2) => {
                                                                                                                                                                        return ((Tuple)IokeObject.dataOf(on2)).elements[index];
                                                                                                                                                                    }));
                                                                                    })));


            obj.RegisterMethod(runtime.NewNativeMethod("Compares this object against the argument. The comparison is only based on the elements inside the tuple, which are in turn compared using <=>.",
                                                       new TypeCheckingNativeMethod("<=>", TypeCheckingArgumentsDefinition.builder()
                                                                                    .ReceiverMustMimic(obj)
                                                                                    .WithRequiredPositional("other")
                                                                                    .Arguments,
                                                                                    (method, on, args, keywords, context, message) => {
                                                                                        object arg = args[0];
                                                                                        var one = ((Tuple)IokeObject.dataOf(on)).elements;
                                                                                        var two = ((Tuple)IokeObject.dataOf(arg)).elements;

                                                                                        int len = System.Math.Min(one.Length, two.Length);
                                                                                        SpaceshipComparator sc = new SpaceshipComparator(context, message);

                                                                                        for(int i = 0; i < len; i++) {
                                                                                            int v = sc.Compare(one[i], two[i]);
                                                                                            if(v != 0) {
                                                                                                return context.runtime.NewNumber(v);
                                                                                            }
                                                                                        }

                                                                                        len = one.Length - two.Length;

                                                                                        if(len == 0) return context.runtime.NewNumber(0);
                                                                                        if(len > 0) return context.runtime.NewNumber(1);
                                                                                        return context.runtime.NewNumber(-1);
                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Returns the arity of this tuple",
                                                       new TypeCheckingNativeMethod.WithNoArguments("arity", runtime.Tuple,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return context.runtime.NewNumber(((Tuple)IokeObject.dataOf(on)).elements.Length);
                                                                                                    })));
            obj.RegisterMethod(runtime.NewNativeMethod("Returns a text inspection of the object",
                                                       new TypeCheckingNativeMethod.WithNoArguments("inspect", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return method.runtime.NewText(Tuple.GetInspect(on));
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Returns a brief text inspection of the object",
                                                       new TypeCheckingNativeMethod.WithNoArguments("notice", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return method.runtime.NewText(Tuple.GetNotice(on));
                                                                                                    })));

            obj.RegisterMethod(runtime.NewNativeMethod("Returns a list representation of this tuple",
                                                       new TypeCheckingNativeMethod.WithNoArguments("asList", obj,
                                                                                                    (method, on, args, keywords, context, message) => {
                                                                                                        return method.runtime.NewList(new SaneArrayList(((Tuple)IokeObject.dataOf(on)).elements));
                                                                                                    })));
        }

        public static string GetInspect(object on) {
            return ((Tuple)(IokeObject.dataOf(on))).Inspect(on);
        }

        public static string GetNotice(object on) {
            return ((Tuple)(IokeObject.dataOf(on))).Notice(on);
        }

        public string Inspect(object obj) {
            StringBuilder sb = new StringBuilder();
            sb.Append("(");
            string sep = "";
            foreach(object o in elements) {
                sb.Append(sep).Append(IokeObject.Inspect(o));
                sep = ", ";
            }
            sb.Append(")");
            return sb.ToString();
        }

        public string Notice(object obj) {
            StringBuilder sb = new StringBuilder();
            sb.Append("(");
            string sep = "";
            foreach(object o in elements) {
                sb.Append(sep).Append(IokeObject.Notice(o));
                sep = ", ";
            }
            sb.Append(")");
            return sb.ToString();
        }
    }
}
